### PR TITLE
Sander vocke trace dump

### PR DESCRIPTION
**Pull request by abadams**
Picked up Sander's PR, rebased and squashed onto master to get rid of all the extraneous commits, and added support for .tmp (so that we can dump things up to 4 dimensions).

### CODE REVIEW

File: util/HalideTraceDump.cpp [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+    exit(1);
+}
+
+int main(int argc, char * const *argv) {
+
+    char *buf_filename = nullptr;
+    char *buf_imagetype = nullptr;
+    BufferOutputOpts outputopts;
+    int c;
+    while ((c = getopt (argc, argv, "i:t:")) != -1) {
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: getopt(

Comment (steven-johnson) [reply util/HalideTraceDump.cpp]: BTW, IIRC 'getopt' isn't available on OSX

File: util/HalideTraceDump.cpp [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+            if (func_info.find(string(p.func())) == func_info.end()) {
+                printf("[INFO] Found Func with tracked accesses: %s\n", p.func());
+                func_info[string(p.func())] = FuncInfo(&p);
+            }
+            func_info[string(p.func())].add_preprocess(&p);
+        }
+    }
+
+    packet_count = 0;
+    fseek(file_desc, 0, SEEK_SET);
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: do we also need to check return value of fseek?

File: util/HalideTraceUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            perror("Failed during read");
+            exit(-1);
+        }
+        return false; //EOF
+    }
+
+    return true;
+}
+
+void bad_type_error(halide_type_t type) {
```
Comment (steven-johnson) [inline util/HalideTraceUtils.cpp]: should this also call exit() or abort()?

File: util/HalideTraceUtils.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+#include "HalideRuntime.h"
+#include <stdio.h>
+
+namespace Halide {
+namespace Internal {
+
+void bad_type_error(halide_type_t type);
+
+// Simple conversion.
+template<typename T>
```
Comment (steven-johnson) [inline util/HalideTraceUtils.h]: Could be done with less casting via halide_scalar_value_t:

```
template<typename T>
T value_as(halide_type_t type, const void* value) {
    const halide_scalar_value_t *v = (const halide_scalar_value_t *)value;
    switch (type.code) {
    case halide_type_int:
        switch (type.bits) {
        case 8:
            return (T) v->i8;
        case 16:
            return (T) v->i16;
        case 32:
            return (T) v->i32;
        case 64:
            return (T) v->i64;
        default:
            bad_type_error(type);
        }
        break;
    case halide_type_uint:
        switch (type.bits) {
        case 1:
            return (T) v->b;
        case 8:
            return (T) v->u8;
        case 16:
            return (T) v->u16;
        case 32:
            return (T) v->u32;
        case 64:
            return (T) v->u64;
        default:
            bad_type_error(type);
        }
        break;
    case halide_type_float:
        switch (type.bits) {
        case 32:
            return (T) v->f32;
        case 64:
            return (T) v->f64;
        default:
            bad_type_error(type);
        }
        break;
    default:
        bad_type_error(type);
    }
    return (T) 0;
}
```

File: util/HalideTraceUtils.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            return (T)(((const int32_t *)value)[0]);
+        case 64:
+            return (T)(((const int64_t *)value)[0]);
+        default:
+            bad_type_error(type);
+        }
+        break;
+    case halide_type_uint:
+        switch (type.bits) {
+        case 8:
```
Comment (steven-johnson) [inline util/HalideTraceUtils.h]: are you deliberately ignoring uint(1) (aka bool)?

Comment (abadams) [reply util/HalideTraceUtils.h]: This is just code from Sander's PR. Probably an oversight - will add it.

File: util/HalideTraceUtils.h (hunk 4 above)
Comment (abadams) [reply util/HalideTraceUtils.h]: Thanks, copy-pasted

File: util/HalideTraceUtils.h (hunk 5 above)
Comment (abadams) [reply util/HalideTraceUtils.h]: (Actually looks like it was my code originally)

File: util/HalideTraceUtils.cpp (hunk 3 above)
Comment (abadams) [reply util/HalideTraceUtils.cpp]: Done

File: util/HalideTraceDump.cpp (hunk 2 above)
Comment (abadams) [reply util/HalideTraceDump.cpp]: I don't think so, looks like it checks ferror immediately below, which will be set if fseek failed

**Comment by abadams** [comment]
builds are clean. ptaal

File: util/HalideTraceDump.cpp [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+            fprintf(stderr, "Negative dimensionality found. Aborting.\n");
+            exit(-1);
+        }
+
+        for (int lane = 0; lane < lanes; lane++) {
+            int coord[16];
+            for (int i = 0; i < dimensions; i++) {
+                coord[i] = p->coordinates()[lanes * i + lane] - min_coords[i];
+            }
+            buf(coord) = p->get_value_as<T>(lane);
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: does it matter than some uninitialized entries in coord[] are passed to buf?

File: util/HalideTraceDump.cpp [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    printf("[INFO] Dumping func '%s' to file: %s\n", name.c_str(), filename.c_str());
+
+    // Rely on save_image to do type-checking
+    Halide::Tools::convert_and_save_image<Buffer<>, check_and_continue>(func.values, filename);
+}
+
+void finish_dump(map<string, FuncInfo> &func_info, BufferOutputOpts output_opts) {
+
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: blank line at start of function is weird

File: util/HalideTraceDump.cpp [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+        // Type
+        printf("      Type: ");
+        if (info.type.code == halide_type_code_t::halide_type_int) {
+            printf("int%d\n", info.type.bits);
+        } else if (info.type.code == halide_type_code_t::halide_type_uint) {
+            printf("uint%d\n", info.type.bits);
+        } else if (info.type.code == halide_type_code_t::halide_type_float) {
+            printf("float%d\n", info.type.bits);
+        }
+        else {
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: } else {

File: util/HalideTraceDump.cpp [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+        "Funcs into individual image files in the current directory.\n"
+        "To generate a suitable binary trace, use Func::trace_stores(), or the\n"
+        "target features trace_stores and trace_realizations, and run with\n"
+        "HL_TRACE_FILE=<filename>.\n";
+    fprintf(stderr, "%s\n", usage.c_str());
+    exit(1);
+}
+
+int main(int argc, char * const *argv) {
+
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: ditto

File: util/HalideTraceDump.cpp [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+                printf("[INFO] Found Func with tracked accesses: %s\n", p.func());
+                func_info[string(p.func())] = FuncInfo(&p);
+            }
+            func_info[string(p.func())].add_preprocess(&p);
+        }
+    }
+
+    packet_count = 0;
+    fseek(file_desc, 0, SEEK_SET);
+    if(ferror(file_desc)){
```
Comment (steven-johnson) [inline util/HalideTraceDump.cpp]: `if (` is used elsewhere

**Comment by steven-johnson** [comment]
LGTM with style nits

File: util/HalideTraceDump.cpp (hunk 6 above)
Comment (abadams) [reply util/HalideTraceDump.cpp]: No, it wants an int array of size dimensions.

**Pull request merged**